### PR TITLE
feat(neo4j): initial adapter

### LINKED ISSUE #113: neo4j adapter.
Hi, has anyone worked on this so far?

I want to try and do it over the next few days.

I'll try and write it in TypeScript although I usually write JavaScript.

If I need some guidance with TS, is a maintainer available to chat? 

For that or any other suggestions, I'm in [Next.js - Discord](https://discord.com/invite/w8p9Psrk2M).

### PR DESCRIPTION
**Pull request by EarthlingDavey**
## Reasoning 💡

> A neo4j adapter for nextauthjs

* Neo4j is in active development and has an active community. 
* [GRANDstack](https://grandstack.io/) can be modified to use next.js as a frontend.
* I am ok to write the database queries, but I will need help with TypeScript. - in fact I have already run into a few errors.

## Checklist 🧢

- [ ] Documentation
- [ ] Tests
- [ ] Ready to be merged

## Affected issues 🎟

closes #113

### CODE REVIEW

**Comment by EarthlingDavey** [comment]
Current status of this PR is:
I am getting TS errors, and cannot run the tests locally yet.

```
Type '(neo4jSession: typeof neo4j.Session) => { getAdapter({ session, secret, ...appOptions }: AppOptions): Promise<{ displayName: string; createUser(profile: Profile & { emailVerified?: Date | undefined; }): Result; ... 13 more ...; deleteVerificationRequest(identifier: string, token: string): Promise<...>; }>; }' is not assignable to type 'Adapter<Session, never, Neo4jUser, Profile & { emailVerified?: Date | undefined; }, Neo4jSession>'.
  Call signature return types '{ getAdapter({ session, secret, ...appOptions }: AppOptions): Promise<{ displayName: string; createUser(profile: Profile & { emailVerified?: Date | undefined; }): Result; getUser(id: string): { ...; }; ... 12 more ...; deleteVerificationRequest(identifier: string, token: string): Promise<...>; }>; }' and '{ getAdapter(appOptions: AppOptions): Promise<AdapterInstance<Neo4jUser, Profile & { emailVerified?: Date | undefined; }, Neo4jSession>>; }' are incompatible.
    The types returned by 'getAdapter(...)' are incompatible between these types.
      Type 'Promise<{ displayName: string; createUser(profile: Profile & { emailVerified?: Date | undefined; }): Result; getUser(id: string): { id: string; }; getUserByEmail(email: string | null): Result; ... 11 more ...; deleteVerificationRequest(identifier: string, token: string): Promise<...>; }>' is not assignable to type 'Promise<AdapterInstance<Neo4jUser, Profile & { emailVerified?: Date | undefined; }, Neo4jSession>>'.
        Type '{ displayName: string; createUser(profile: Profile & { emailVerified?: Date | undefined; }): Result; getUser(id: string): { id: string; }; getUserByEmail(email: string | null): Result; ... 11 more ...; deleteVerificationRequest(identifier: string, token: string): Promise<...>; }' is not assignable to type 'AdapterInstance<Neo4jUser, Profile & { emailVerified?: Date | undefined; }, Neo4jSession>'.
          The types of 'createUser(...).then' are incompatible between these types.
            Type '<TResult1 = QueryResult, TResult2 = never>(onFulfilled?: ((value: QueryResult) => TResult1 | PromiseLike<TResult1>) | null | undefined, onRejected?: ((reason: any) => TResult2 | PromiseLike<...>) | ... 1 more ... | undefined) => Promise<...>' is not assignable to type '<TResult1 = Neo4jUser, TResult2 = never>(onfulfilled?: ((value: Neo4jUser) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | null | undefined) => Promise<...>'.
              Types of parameters 'onFulfilled' and 'onfulfilled' are incompatible.
                Types of parameters 'value' and 'value' are incompatible.
                  Type 'QueryResult' has no properties in common with type 'Neo4jUser'.
```

This is my first TS project so any assistance with debugging / fixing the error is appreciated.

**Comment by balazsorban44** [comment]
Thanks for opening this! I'll have a look and try to help, hopefully soon. It looks like a good start!

**Comment by EarthlingDavey** [comment]
Hi @balazsorban44 thats awesome. I look forward to getting it done. Thank you :)

**Comment by EarthlingDavey** [comment]
I'm hoping to keep some momentum with this PR.

@balazsorban44 if you are tight on time - could you add `help wanted` or maybe `good first issue` labels. 
Do you think someone else with junior TS experience can fix my error?

Alternatively, I can work on a JS version and port to TS later. But, this is not really preferred for me 😅

**Comment by balazsorban44** [comment]
@ndom91 might be able to help out. 😊

**Comment by ndom91** [comment]
@EarthlingDavey I see you already have some tests / testing scripts in place! The first step would be to add the neo4j package subdirectory to the `test-adapters` matrix [here](https://github.com/usedandloved/nextauthjs-adapters/blob/f0ff3e85b8be1cebe713a517a3cbdf720702716f/.github/workflows/main.yml#L36).

Then ci will run tests on this PR as well and we can work from there :+1:

**Comment by EarthlingDavey** [comment]
@balazsorban44 thanks. 
@ndom91 sure, thanks for the advice - it's done.

**Comment by ndom91** [comment]
@EarthlingDavey the neo4j directory also needs to be added to `.github/workflows/canary.yml`. Then the tests shoudl run on this PR!

**Comment by EarthlingDavey** [comment]
Hi @ndom91 done... I missed your PR 6 days ago 🤦‍♂️
I know I already mentioned I'm on the nextjs discord... but I'll say again because I think I could bash through this really quickly with a bit of realtime guidance.

**Comment by ndom91** [comment]
I can take a look with you if you want. Want to pair on it? 

Our primary maintainer, Balazs, is on holiday currently so I don't want to bug him.

**Comment by EarthlingDavey** [comment]
> Balazs, is on holiday 

I totally respect that decision 👍 

Pair, or whatever time/guidance you can afford. I can do some time today and Mon - Fri about 7 - 3pm GMT

**Comment by ndom91** [comment]
@EarthlingDavey okay great, i'm available thsi evening (GMT+2:00) starting at like 17-18 and during the week any day during the evening as well (18 -> @ GMT+2:00).

Looks like the test finally ran, but failed. So we definitely have some work to do :)

**Comment by EarthlingDavey** [comment]
@ndom91 can we go with tomorrow 19:00 @ GMT + 2? For about a couple of hours?

**Comment by ndom91** [comment]
@EarthlingDavey yup sounds good to me! Feel free to send me an invite to yo@ndo.dev.

**Comment by EarthlingDavey** [comment]
@ndom91 please can you run tests. They are passing locally :) except for an error that I need some help with

`Jest has detected the following 1 open handle potentially keeping Jest from exiting:`

**Comment by ndom91** [comment]
Hey @lluia so I did a quick video call with @EarthlingDavey and we got the neo4j tests up and running, however there were still some open questions about types here. Can you give this a look whenever you have some time?

**Comment by EarthlingDavey** [comment]
Thanks @ndom91 :) @lluia I'll post the questions tomorrow. For now they are meaked `TODO: ` in `/packages/neo4j`

**Comment by balazsorban44** [comment]
As a heads up here, I found out that our adapter TS could be simplified even more, so expect a PR from me on it in the core repo! it might be out in a few days, which would maybe help or even eliminate the work needed here to fix TS issues.

**Comment by EarthlingDavey** [comment]
Hi folks, sorry for the delay with this.
I've refetched from upstream. 
If the tests pass and it looks ok from your end would you mind merging into canary?
If it's passing, I'll address my queries later as I don't think the answers will introduce any breaking changes.
Thanks :)

**Comment by roelandmoors** [comment]
I tried to use this, but I get an error when trying to login with a Google account.
But this is my first try with next-auth. Not sure if this is a problem with the neo4j adapter or just the Google provider?
There is a warning on the Google provider [page](https://next-auth.js.org/providers/google):
`Google only provides Refresh Token to an application the first time a user signs in.`
But this was the first time. Not sure what is causing this?


```
[next-auth][error][oauth_callback_handler_error]
https://next-auth.js.org/errors#oauth_callback_handler_error Neo4jError: Expected parameter(s): refreshToken

    at captureStacktrace (/Users/foo/project/node_modules/neo4j-driver-core/lib/result.js:239:17)
    at new Result (/Users/foo/project/node_modules/neo4j-driver-core/lib/result.js:59:23)
    at newCompletedResult (/Users/foo/project/node_modules/neo4j-driver-core/lib/transaction.js:372:12)
    at Object.run (/Users/foo/project/node_modules/neo4j-driver-core/lib/transaction.js:226:20)
    at Transaction.run (/Users/foo/project/node_modules/neo4j-driver-core/lib/transaction.js:98:34)
    at eval (webpack-internal:///../nextauthjs-adapters/packages/neo4j/dist/account.js:20:19)
    at TransactionExecutor._safeExecuteTransactionWork (/Users/foo/project/node_modules/neo4j-driver-core/lib/internal/transaction-executor.js:92:26)
    at TransactionExecutor._executeTransactionInsidePromise (/Users/foo/project/node_modules/neo4j-driver-core/lib/internal/transaction-executor.js:83:34)
    at /Users/foo/project/node_modules/neo4j-driver-core/lib/internal/transaction-executor.js:40:19
    at new Promise (<anonymous>) {
  name: 'LinkAccountError'
```
}

PS: in the README are some references to pouchdb. I guess that needs an update.

**Comment by EarthlingDavey** [comment]
Hi @roelandmoors, thanks for letting me know.

I've replicated your error. I can see that the line before your error message is like this:

```
[next-auth][debug][neo4j__link_account] 
8d88a0dc-8c46-4c84-8b86-605264ba0467 
google 
oauth 
redacted
undefined <-- refresh token is undefined
redacted
null
```

It's strange because the [google provider docs](https://next-auth.js.org/providers/google) say it should be included the first time a user signs in.

I've fixed the relevant code in my provider. It should work now.

**Comment by EarthlingDavey** [comment]
Plugin maintainers, I have a couple of questions.

1. Do you know if this looks ok? It seems the docs might be wrong. As google didn't provide a refresh token even on first sign in.
1. Should I go ahead and add a test for the above case to the basic-tests?
1. On a tangent, I tested twitter integration without requesting an email [docs](https://next-auth.js.org/providers/twitter). And my adapter failed because I didn't test for this. Is this also a test that I should add to basic-tests?

**Comment by roelandmoors** [comment]
I tried again with the latest source and now it seems to work fine.
The user/account/session is created and I'm logged in.
image

**Comment by ndom91** [comment]
> Plugin maintainers, I have a couple of questions.
> 
> 1. Do you know if this looks ok? It seems the docs might be wrong. As google didn't provide a refresh token even on first sign in.
> 2. Should I go ahead and add a test for the above case to the basic-tests?
> 3. On a tangent, I tested twitter integration without requesting an email [docs](https://next-auth.js.org/providers/twitter). And my adapter failed because I didn't test for this. Is this also a test that I should add to basic-tests?

Hey @EarthlingDavey first of all, great work keeping this up-to-date - thanks! Unfortunately I'm not too well versed in the oauth side of things, maybe Balazs has more insight into the Google behavior. He's currently on holiday, however for another 1.5-2 weeks I believe. 

The most important thing I want to mention though, is that he is significantly simplifying the adapter api, including taking functions like `hashToken` into the core of next-auth so each adapter doesn't have to reimplement it themselves, as well as simplifying the typescript types. Goal of all of this being that adapter authors (like yourself) really only have to translate the next-auth actions / functions into the specific database calls for whichever adapter.

Long story short, maybe I'd hold off on finishing this up until that is done on his part.

I believe the two relevant PRs are:
`next-auth`: https://github.com/nextauthjs/next-auth/pull/2361
`adapters`: Example for migrating prisma to the new api: https://github.com/nextauthjs/adapters/pull/171

**Comment by EarthlingDavey** [comment]
@ndom91 cheers for taking a look.

Sure no problem, I'll wait a few weeks. Then make amends to work with the new API.

In the mean time I'll try to support what I've done so far. ( @roelandmoors - let me now how you get on, I'll be using this in prod in 6 weeks so it's all appreciated 👍 )

**Comment by balazsorban44** [comment]
@EarthlingDavey  Hopefully you are still around, the new API has been merged, and you can see Prisma and the Fauna adapter implementations for reference!

https://github.com/nextauthjs/adapters/blob/next/packages/prisma/src/index.ts
https://github.com/nextauthjs/adapters/blob/next/packages/fauna/src/index.ts

Please base your branch against `next` (the easiest way might be to close this PR and open a new one where you start off with the latest `next` to be honest), and do the necessary changes.

**Comment by EarthlingDavey** [comment]
@balazsorban44 Hi, I'm still here. Thanks for your work and links to reference.

I'll work on this soon and follow your suggestion.

**Comment by balazsorban44** [comment]
Awesome! Happy to help!
